Make sure queries by name to Folder/Label/Category account for truncation

jstejada opened the pull request:
- Cleans up Folder/Label/Category name sanitization code which was mostly duplicated
- Added a custom string type for Category names to make sure we always store properly sanitized strings, and extends `__eq__` comparison to compare against the truncated string, which is what we actually store in the database, even if the user queries using a longer (original) name. Going forward, any queries by name will now compare against the correct string.
- This only affects querying-- we still use the full folder name to be able to select the folder correctly in IMAP inside the sync engine
- Improves logging -- don't interpolate strings in the logs!

- This addresses most occurrences of https://sentry.nylas.com/sentry/sync-prod/group/8243/events/

bengotow commented on inbox/mailsync/backends/imap/generic.py:
```diff
@@ -118,8 +119,10 @@ def __init__(self, account_id, namespace_id, folder_name,
+                    .rstrip()[:MAX_FOLDER_NAME_LENGTH]
                 folder = db_session.query(Folder). \
-                    filter(Folder.name == folder_name,
+                    filter(Folder.name == folder_stored_name,
```
Oh nice catch—I wonder if there's a way we could perform this mapping lower down in the Folder model so we don't have to remember to do it all the time? Would be nice if we could say "limit the length of the comparator in the where clause to the max length of the field"...

spang replied:
Hmmmm or at least put it as a method on the Folder model or something like that. Generally in Python you can override comparator behaviour by defining a __cmp__ method on the object, but I'm not sure how nice that will play with SQLAlchemy's declarative declarations here.

jstejada replied:
good call

jstejada commented on inbox/models/category.py:
```diff
@@ -39,21 +48,29 @@ def API_OBJECT_NAME(self):
+    def sanitize_name(cls, name):
+        # g_label may not have unicode type (in particular for a numeric
+        # label, e.g. '42'), so coerce to unicode.
+        name = unicode(name)
```
this line was the only difference in sanitizing between Label and Folder, I /thinl/ there would be no harm in also coercing to unicode for Folder names, but I might be mistaken

spang replied:
I think this should be fine. I am in favour of having all of our internal processing use unicode objects and only convert strings to bytestrings when needed to transmit them to the outside world.

spang commented on inbox/models/category.py:
```diff
@@ -18,6 +19,14 @@
 EPOCH = datetime.utcfromtimestamp(0)
 
 
+class CategoryNameType(String):
```
Can you put a docstring on this describing what it's for? I think this would make it more clear to future readers!

jstejada replied:
definitely @spang

jstejada commented on inbox/models/folder.py:
```diff
@@ -36,9 +36,9 @@ class Folder(MailSyncBase, UpdatedAtMixin, DeletedAtMixin):
-    canonical_name = Column(String(MAX_FOLDER_NAME_LENGTH), nullable=False,
+    name = Column(CategoryNameType(MAX_INDEXABLE_LENGTH,
+                  collation='utf8mb4_bin'), nullable=False)
+    canonical_name = Column(String(MAX_INDEXABLE_LENGTH), nullable=False,
```
It seems that `canonical_name` for Folder/Label and `name` for Category are never truncated, probably due to the fact that they corresponds to the `role`, which can never be that long. I wasn't sure if adding truncation to these fields was necessary or would cause any trouble, and given that we never have truncated it I left it as so.

spang commented on inbox/models/category.py:
```diff
@@ -39,21 +48,29 @@ def API_OBJECT_NAME(self):
     name = Column(String(MAX_INDEXABLE_LENGTH), nullable=False, default='')
-    display_name = Column(String(MAX_INDEXABLE_LENGTH,
-                                 collation='utf8mb4_bin'), nullable=False)
+    display_name = Column(CategoryNameType(MAX_INDEXABLE_LENGTH,
```
If you wanted to make this even simpler, you could override `__init__`  too and pass in `MAX_INDEXABLE_LENGTH` from within `CategoryNameType()`'s constructor to the superclass—that would eliminate any chance of accidentally using a different length anywhere.

spang commented:
This diff LGTM for the most part—thanks for cleaning things up! I left a few minor comments inline, and I would also love to see a unit test for the custom comparator type.

bengotow commented on inbox/models/category.py:
```diff
@@ -39,21 +60,28 @@ def API_OBJECT_NAME(self):
     name = Column(String(MAX_INDEXABLE_LENGTH), nullable=False, default='')
-    display_name = Column(String(MAX_INDEXABLE_LENGTH,
-                                 collation='utf8mb4_bin'), nullable=False)
+    display_name = Column(CategoryNameString(), nullable=False)
```
This is really nice, but it's confusing that the new column type pre-processes values you want to compare it against, but does not pre-process values you actually assign to it. I could still say: "display_name = super-long-string", and then display_name == super-long-string would be false.

I think it'd be really cool to encapsulate both the value coercion and comparator coercion into the same type class, so we don't need to have a separate block of code below `@validates("display_name")`.

I /think/ this might work:

```
class CategoryNameString(String):
   __init__: (transformCb) =>
      this.transformCb = transformCb

   class comparator_factory(String.Comparator):
       def __eq__(self, other):
             other = Category.sanitize_name(other)
             return self.operate(operators.eq, other)

    def bind_processor(self, dialect):
        def process(value):
            return this.transformCb(super.bind_processor(dialect)(value))
        return process

```

If we could encapsulate both, we could make this more generically useful:

```
display_name = Column(StringWithTransform(MAX_INDEXABLE_LENGTH, transform=self. sanitize_name), nullable=False)

```

jstejada commented:
@spang, @bengotow, addressed your comments and made the custom string type more generic as per discussion with ben -- it might be worth reusing for stuff like filename truncation. Also added some unit tests. Could you take a look at this latest diff? It seems that the bot didn't catch spang's lgtm.

bengotow commented on inbox/models/category.py:
```diff
@@ -6,6 +6,7 @@
 from sqlalchemy.orm.exc import MultipleResultsFound
 from sqlalchemy.ext.hybrid import hybrid_property
 
+from inbox.sqlalchemy_ext.util import StringWithTransform
```
<3

bengotow commented on inbox/models/category.py:
```diff
@@ -18,6 +19,36 @@
+def sanitize_name(name):
+    # g_label may not have unicode type (in particular for a numeric
+    # label, e.g. '42'), so coerce to unicode.
+    name = unicode(name)
```
I believe that doing `unicode(unicode(name))` fails, so we may need to wrap this in a check to make sure `name` is not already `isinstance(name, unicode)`

bengotow commented on inbox/models/category.py:
```diff
@@ -18,6 +19,36 @@
+    return name
+
+
+class CategoryNameString(StringWithTransform):
```
I wonder if this extra subclass is still useful? I actually feel like having the StringWithTransform(Sanitize_name) followed by the sanitize_name validation in the same file might be easiest to understand? Totally up to you!

bengotow commented:
This is really nice. Thanks for cleaning up the duplicate logic! I think in the future we should definitely consider using `StringWithTransform` for filenames and other things. Doing a quick grep for `[:` it looks like there are several other places where we're guarding against string length that might be better placed into ORM-level transforms 👍 

Ship it! LGTM

spang commented:
I like it! looking forward to more changes that make the codebase easier to understand and work with and it harder for us to introduce bugs :) :+1:
